Dataflow refining and inotify implementation

tjk584 opened the pull request:
## Overview
This pull request changes the data flow repository in a number of ways:
- Implements inotify to trigger dataflow scripts, replacing crontab
- Re-organizes the repo and re-factors much of the code
- Moves telemetry data from email to staging for the Engineering Dashboard

## Changes

#### Organizational changes:
- Added superdarn-cssdp directory for all scripts running on that computer
- Moved auto_borealis_share from USASKmirror-Scripts repository here (in archived scripts)
- Created inotify_daemons directory for the inotifywait daemons that run each script
   - `systemd` services copied here for reference also 
- Copied all old dataflow scripts to archive directory for reference 
- Renamed dataflow scripts to be more concise and clear

#### General changes:
- Made variable declaration and code syntax consistent across all scripts
- Cleaned up code, changed deprecated bash syntax
- Changed logging for all scripts to create daily logs. Added more context to logging information
- Added creation of summary logs for each script. Summary logs only contain information on the status of each file operation (if file transfer/conversion was successful or not)
- Eliminated all emailing of script errors, summary logs are instead synched to Chapman for telemetry uses
- Moved functions used by dataflow scripts to a separate bash file
- Created a function to correctly check file transfer status for both remote and local transfers using `md5sum`
- Created a `config.sh` file that contains all information needed to make basic changes to the dataflow. This includes specifying which sites are operating via the NAS, what files each site is converting, and which sites have memory/bandwidth limitations
   - The data flow scripts handle these cases and automatically switch to operate correctly. I.E. if the NAS is down for a site, the scripts will know to rsync only rawacf files to the site-linux computer instead of the NAS

#### Inotify changes:
- inotify daemon scripts for each data flow stage were made, available in /inotify_daemons. These scripts run as a daemon via `systemd`, and trigger in the following order:
   - borealis.daemon triggers when Borealis finished writing a 2-hour file
   - site-linux.daemon triggers when borealis.daemon finishes
   - campus.daemon triggers when site-linux.daemon finishes
- Each daemon script executes all data flow scripts for their respective computer sequentially
   - All data flow scripts operate as normal, and can be used outside of the inotify daemons
   - More info on how inotify works is available in the `README.md` and the associated trello card
- rsync_to_nas was modified to accept a borealis file as a parameter for filtering purposes. If no parameter is given, it operates as it did before these changes

#### Telemetry:
- All dataflow scripts synchronize their summary logfiles to telemetry@chapman.usask.ca
- The summary logfiles contain various info about the state of the data flow, and the state of individual operations done by each script
- A python script parses these log files and compiles telemetry data, which is then displayed on the Engineering dashboard.

## Testing
This branch has been running successfully on each of the sites for the last week.

## Extra information
More details are given in the associated Trello cards: 
https://trello.com/c/74fgFOza/67-implement-inotify-for-site-to-campus-dataflow 
https://trello.com/c/rOSTPg5Q/69-refine-dataflow-scripts

kevinkrieger commented on superdarn-cssdp/convert_on_campus:
```diff
@@ -0,0 +1,141 @@
+        printf "\nConverting the following files:\n"
+        printf '%s\n' "${RAWACF_CONVERT_FILES[@]}"
+    else
+        printf "\nNo files to be converted.\n"
```
Would this be an error - considering this script should run only if there's a flag generated?

tjk584 replied:
I think there could be a case where files fail conversion or transfer earlier along in the dataflow, but the inotify flags continue to trigger subsequent scripts. It would be good to log and notify the engineering dashboard in this case, however

kevinkrieger commented:
I think it just needs to lock scripts to only run one at a time and it should be good

tjk584 commented on borealis/rsync_to_nas:
```diff
@@ -0,0 +1,144 @@
+
+# Threshold (in minutes) for selecting files to transfer with `find`.
+# Set to filter out files currently being written to by Borealis.
+readonly FILE_THRESHOLD=0.1		# 0.1 min = 6 s
```
What sort of idea do you have in mind? I could implement the 6 s delay outside this script within the inotify daemon, but I'm not sure how to use inotify to select files not currently being written to

kevinkrieger replied:
Hmm maybe we should discuss in person

tjk584 commented on inotify_daemons/rsync_to_campus.daemon:
```diff
@@ -0,0 +1,22 @@
+inotifywait -r --monitor --syslog --timefmt '%Y%m%d.%H:%M:%S' --format '%T %w %e %f' -e MOVED_TO $WATCH_DIR | \
+while read TIME DIRECTORY ACTION FILE # These must line up with format parameters
+do
+    if [[ "$FILE" =~ \.convert_and_restructure_flag$ ]]   # Only interested in 2 hour files
```
I believe it symbolized the end of a line in regular expressions - I don't think its crucial in this case though

tjk584 commented on superdarn-cssdp/convert_on_campus:
```diff
@@ -0,0 +1,141 @@
+set -o pipefail  # don't hide errors within pipes
+
+# source the RADAR_ID, SDCOPY and other things
+source "${HOME}/.bashrc"
```
I don't think its used in this script, correct.

tjk584 commented on superdarn-cssdp/convert_on_campus:
```diff
@@ -0,0 +1,141 @@
+printf "Executing $(basename "$0") on $(hostname)\n" | tee --append $SUMMARY_FILE
+date --utc "+%Y%m%d %H:%M:%S UTC" | tee --append $SUMMARY_FILE
+
+if [[ " ${SITES[*]} " =~ " ${RADAR_ID} " ]]; then
```
Yeah I think I should be able to implement something within the inotify daemon scripts similar to the `flock` currently used in the crontabs

tjk584 commented:
Changed rsync_to_nas's filtering to be time-agnostic. The script now filters based on the filename of the script that triggered inotify, instead of waiting for 6 s before taking all older files. This will prevent future problems if Borealis somehow doesn't write to file within the first 6 seconds of a 2-hour block

tjk584 commented:
Changed inotify daemon setup: Instead of having individual inotify daemon scripts for each dataflow script, I changed it so each computer has its own inotify daemon script, and all scripts executed on the same computer are triggered by the same daemon. For example, on the site-linux computer, the site-linux daemon executes `convert_and_restructure` and `rsync_to_campus` one after the other within the same script. The single inotify daemon then only has to trigger inotify daemons on other computers, not on the same computer. 

Additionally, I moved all inotify flag handling and references from the data flow scripts to the inotify daemon scripts. This reduces the coupling and complexity of the data flow scripts and allows them to be run on their own without any use of inotify.

kevinkrieger commented on inotify_daemons/distribute_borealis_data.daemon:
```diff
@@ -21,6 +22,8 @@ do
-        $HOME/data_flow/superdarn-cssdp/distribute_borealis_data $RADAR_ID
+        touch $LOCK_FILE
+        # Lock with radar specific file to ensure only one instance of script runs for each site
+        flock -n --verbose $LOCK_FILE $HOME/data_flow/superdarn-cssdp/distribute_borealis_data $RADAR_ID
```
same here, -c infront of command

kevinkrieger reviewed (approved):
looks good

RemingtonRohel commented on inotify_daemons/borealis.daemon:
```diff
@@ -0,0 +1,69 @@
+
+###################################################################################################
+
+exec &>> $LOGFILE
```
What does this line do?

tjk584 replied:
It redirects STDOUT and STDERR to $LOGFILE for all commands following. Useful for directing log output within a bash script

RemingtonRohel commented on inotify_daemons/borealis.daemon:
```diff
@@ -0,0 +1,69 @@
+readonly FLAG_DEST="/home/transfer/test_data_flow/data_flow/.inotify_watchdir" # For SAS testing
+
+# Flag to send to start next script
+readonly FLAG_OUT="${HOME}/test_data_flow/data_flow/.inotify_flags/.borealis_flag" # For SAS testing
```
These should be changed for the PR since they are testing values, yes?

tjk584 replied:
Correct, these will be changed once SAS testing is finished

RemingtonRohel commented on inotify_daemons/borealis.daemon:
```diff
@@ -0,0 +1,69 @@
+            echo "$(date --utc "+%Y%m%d %H:%M:%S UTC") - Inotify triggered on $(basename $0)"
+
+            # Pass current filename so script knows what to omit
+            $HOME/test_data_flow/data_flow/borealis/rsync_to_nas $FILE # For SAS testing
```
Another instance of test directory being used.

RemingtonRohel commented on inotify_daemons/campus.daemon:
```diff
@@ -0,0 +1,58 @@
+# Author: Theodore Kolkman
+#
+# Inotify script to be run on SuperDARN-CSSDP. Executes convert_on_campus and 
+# distribute_borealis_data, one after the other#
```
```suggestion
# distribute_borealis_data, one after the other
```

RemingtonRohel commented on inotify_daemons/campus.daemon:
```diff
@@ -0,0 +1,58 @@
+readonly FLAG_DEST="${HOME}/test_data_flow/data_flow/.inotify_watchdir" # For SAS testing
+
+# Flag sent out to trigger next dataflow script (TODO)
+readonly FLAG_OUT="${HOME}/test_data_flow/data_flow/.inotify_flags/.campus_dataflow_${RADAR_ID}" # For SAS testing
```
Test directories, change for PR?

RemingtonRohel commented on inotify_daemons/campus.daemon:
```diff
@@ -0,0 +1,58 @@
+        # Reset flag by removing .site-linux_flag_RADARID, so inotify can be triggered again
+        rm --verbose --force "${DIRECTORY}${FILE}"
+
+        # Send out "flag" to trigger next script with inotify (TODO)
```
What is still TODO with this? It looks like the flag is being written

tjk584 replied:
The TODO is for data flow scripts beyond the ones within this repository to use this flag. I'll clarify this in the TODO comment

RemingtonRohel commented on inotify_daemons/site-linux.daemon:
```diff
@@ -0,0 +1,53 @@
+script_name=$(echo $(basename $0) | cut --fields 1 --delimiter '.')
+readonly LOGFILE="${HOME}/logs/inotify_daemons/${script_name}.log"
+
+# Location of inotify watch directory for flags on superdar-cssdp
```
```suggestion
# Location of inotify watch directory for flags on superdarn-cssdp
```

RemingtonRohel commented on inotify_daemons/site-linux.daemon:
```diff
@@ -0,0 +1,53 @@
+readonly FLAG_DEST="/home/mrcopy/test_data_flow/data_flow/.inotify_watchdir" # For SAS testing
+
+# Flag sent out to trigger superdarn-cssdp dataflow
+readonly FLAG_OUT="${HOME}/test_data_flow/data_flow/.inotify_flags/.site-linux_flag_${RADARID}" # For SAS testing
```
Some more testing directories to be changed?

RemingtonRohel commented on inotify_daemons/site-linux.daemon:
```diff
@@ -0,0 +1,53 @@
+        echo "$(date --utc "+%Y%m%d %H:%M:%S UTC") - Inotify triggered on $(basename $0)"
+
+        echo "Executing convert_and_restructure..."
+        $HOME/test_data_flow/data_flow/site-linux/convert_and_restructure # For SAS testing
+        echo "Executing rsync_to_campus..."
+        $HOME/test_data_flow/data_flow/site-linux/rsync_to_campus # For SAS testing
```
More test directories

RemingtonRohel commented on library/data_flow_functions.sh:
```diff
@@ -0,0 +1,66 @@
+	# Argument 1 should be the subject
+	# Argument 2 should be the body
+	# What email address to send to?
+	EMAILADDRESS="kevin.krieger@usask.ca"
```
Probably a good idea to add some more emails to this list, redundancy with notifications never hurts.

tjk584 replied:
I removed all uses of this email function from the code actually, I could probably delete this function. The plan is to send summary logfiles to the Engineering dashboard and parse them there for any errors or notifications we need

RemingtonRohel replied:
Sounds good. I support deleting the function.

RemingtonRohel commented on library/data_flow_functions.sh:
```diff
@@ -0,0 +1,66 @@
+# the source file. If the destination file is on a different computer, specify
+# the ssh address of the destination computer (Ex. radar@10.65.0.32)
+# Argument 1: Source file
+# Argument 2: Destination file (if on remote computer, specify in argument 3)
```
Slightly confused by this, can you add an example usage in this comment for both use cases?

tjk584 replied:
For verifying on transfers within the same computer (or to a NAS), you just need the absolute path to both files you are comparing. This would be for all local transfers (non-ssh transfers), including transfers between Borealis and the NAS. For example:

```
verify_transfer /data/borealis_data/20221021.1600.00.sas.0.rawacf.hdf5.site /borealis_nfs/borealis_data/backup/20221021.1600.00.sas.0.rawacf.hdf5.site
```

For verifying transfer to a remote computer, you also need to specify the ssh user and address (i.e. radar@10.65.0.32) as well as the absolute path to the file on the remote machine. For example, when transferring from the site NAS to campus:

```
verify_transfer /borealis_nfs/borealis_data/rawacf_array/20221021.1600.00.sas.0.rawacf.hdf5 /sddata/inotify_data_flow/sas_data/20221021.1600.00.sas.0.rawacf.hdf5 mrcopy@128.233.224.39
```

I'll add these examples to the function documentation as well to make it clearer

RemingtonRohel replied:
Thanks, that clears it up. I was confused by the argument descriptions in the comment, it sounded like you should specify the destination file as the third argument if you were transferring to a remote computer. With this example I see that I was just reading it wrong

RemingtonRohel commented on inotify_daemons/borealis.daemon:
```diff
@@ -0,0 +1,72 @@
+
+# Location of inotify watch directory for flags on site linux. Should match up with site-linux
+# inotify's WATCH_DIR
+readonly REMOTE_FLAG_DEST="/home/transfer/test_data_flow/data_flow/.inotify_watchdir" # For SAS testing
+
+# Flag to send to start next script
+readonly FLAG_SRC="${HOME}/test_data_flow/data_flow/.inotify_flags" # For SAS testing
```
Couple of testing directories here

RemingtonRohel commented on inotify_daemons/campus.daemon:
```diff
@@ -0,0 +1,61 @@
+
+RADAR_ID=$1
+
+readonly WATCH_DIR="${HOME}/test_data_flow/data_flow/.inotify_watchdir" # For SAS testing
+mkdir -p $WATCH_DIR
+readonly LOCK_FILE="${HOME}/test_data_flow/data_flow/.${RADAR_ID}_dataflow_lock" # For SAS testing
```
Couple of testing directories

RemingtonRohel commented on inotify_daemons/campus.daemon:
```diff
@@ -0,0 +1,61 @@
+readonly LOGFILE="${HOME}/logs/inotify_daemons/${script_name}_${RADAR_ID}.log"
+
+# Location of inotify watch directory for flags. Should match up with next scripts WATCH_DIR (TODO)
+readonly REMOTE_FLAG_DEST="${HOME}/test_data_flow/data_flow/.inotify_watchdir" # For SAS testing
```
Test directory

RemingtonRohel commented on inotify_daemons/campus.daemon:
```diff
@@ -0,0 +1,61 @@
+readonly REMOTE_FLAG_DEST="${HOME}/test_data_flow/data_flow/.inotify_watchdir" # For SAS testing
+
+# Flag sent out to trigger next dataflow script (TODO)
+readonly FLAG_SRC="${HOME}/test_data_flow/data_flow/.inotify_flags" # For SAS testing
+mkdir -p $FLAG_SRC
+readonly FLAG_OUT="${FLAG_SRC}/.campus_dataflow_flag_${RADAR_ID}" # For SAS testing
```
Testing directories

RemingtonRohel commented on inotify_daemons/campus.daemon:
```diff
@@ -0,0 +1,61 @@
+
+        # Lock with radar specific file to ensure only one instance of script runs for each site
+        echo "Executing convert_on_campus..."
+        flock -n $LOCK_FILE $HOME/test_data_flow/data_flow/superdarn-cssdp/convert_on_campus $RADAR_ID # For SAS testing
+        echo "Executing distribute_borealis_data..."
+        flock -n $LOCK_FILE $HOME/test_data_flow/data_flow/superdarn-cssdp/distribute_borealis_data $RADAR_ID # For SAS testing
```
Test directories

RemingtonRohel commented on inotify_daemons/site-linux.daemon:
```diff
@@ -0,0 +1,57 @@
+
+source "${HOME}/.bashrc" # source the RADARID and SDCOPY
+
+readonly WATCH_DIR="${HOME}/test_data_flow/data_flow/.inotify_watchdir" # For SAS testing
```
Test directory

RemingtonRohel commented on inotify_daemons/site-linux.daemon:
```diff
@@ -0,0 +1,57 @@
+script_name=$(echo $(basename $0) | cut --fields 1 --delimiter '.')
+readonly LOGFILE="${HOME}/logs/inotify_daemons/${script_name}.log"
+
+# Location of inotify watch directory for flags on superdar-cssdp. Should match up with 
```
```suggestion
# Location of inotify watch directory for flags on superdarn-cssdp. Should match up with 
```

RemingtonRohel commented on inotify_daemons/site-linux.daemon:
```diff
@@ -0,0 +1,57 @@
+
+# Location of inotify watch directory for flags on superdar-cssdp. Should match up with 
+# superdarn-cssdp inotify's WATCH_DIR
+readonly REMOTE_FLAG_DEST="/home/mrcopy/test_data_flow/data_flow/.inotify_watchdir" # For SAS testing
+
+# Flag sent out to trigger superdarn-cssdp dataflow
+readonly FLAG_SRC="${HOME}/test_data_flow/data_flow/.inotify_flags" # For SAS testing
+mkdir -p $FLAG_SRC
```
Test directories

RemingtonRohel commented on borealis/rsync_to_nas:
```diff
@@ -0,0 +1,142 @@
+
+# Directory the files will be transferred to
+if [[ "$TRANSFER_TO_NAS" == true ]]; then
+	readonly DEST="/borealis_nfs/testing/inotify_data_flow/daily/" # On NAS	# For SAS testing
```
Test directory, need to change it for the PR?

RemingtonRohel commented on borealis/rsync_to_nas:
```diff
@@ -0,0 +1,142 @@
+
+###################################################################################################
+
+source "${HOME}/test_data_flow/data_flow/library/data_flow_functions.sh" # For SAS testing
```
Test directory, need to change it for the PR?

RemingtonRohel commented on site-linux/borealis_convert_file.py:
```diff
@@ -25,6 +25,7 @@
 import datetime
 import os
 import sys
+from tracemalloc import start
```
What is this used for? I don't see start used in the changes you made.

tjk584 replied:
Not sure, might have been a mistake. I'll delete it

RemingtonRohel commented on site-linux/convert_and_restructure:
```diff
@@ -0,0 +1,205 @@
+###################################################################################################
+
+source "${HOME}/.bashrc" # source the RADARID, SDCOPY and other things
+source "${HOME}/test_data_flow/data_flow/library/data_flow_functions.sh" # Load in function library # For SAS testing
```
Test directory

RemingtonRohel commented on site-linux/convert_and_restructure:
```diff
@@ -0,0 +1,205 @@
+###################################################################################################
+
+# Define directories
+readonly DATA_DIR="/borealis_nfs/testing/inotify_data_flow" # For SAS testing
```
Test directory

RemingtonRohel commented on site-linux/convert_and_restructure:
```diff
@@ -0,0 +1,205 @@
+    if [[ " ${RAWACF_SITES[*]} " =~ " ${RADARID} " ]]; then
+        printf "\nConverting ${f}\n"
+        printf "python3 remove_record.py $(basename ${f})\n"
+        output=$(python3 ${HOME}/test_data_flow/data_flow/site-linux/remove_record.py ${f}) # For SAS testing
```
Test directory

RemingtonRohel commented on site-linux/convert_and_restructure:
```diff
@@ -0,0 +1,205 @@
+            printf "Removed records from ${f}:\n${remove_records_output}\n" | tee --append $SUMMARY_FILE
+        fi
+        printf "python3 borealis_convert_file.py --dmap $(basename ${f})\n"
+        python3 "${HOME}/test_data_flow/data_flow/site-linux/borealis_convert_file.py" --dmap $f # For SAS testing
```
Test directory

RemingtonRohel commented on site-linux/convert_and_restructure:
```diff
@@ -0,0 +1,205 @@
+    if [[ " ${BFIQ_SITES[*]} " =~ " ${RADARID} " ]]; then
+        printf "\nConverting ${f}\n"
+        printf "python3 remove_record.py $(basename ${f})\n"
+        output=$(python3 ${HOME}/test_data_flow/data_flow/site-linux/remove_record.py ${f}) # For SAS testing
```
Test dir

RemingtonRohel commented on site-linux/convert_and_restructure:
```diff
@@ -0,0 +1,205 @@
+            printf "Removed records from ${f}:\n${remove_records_output}\n" | tee --append $SUMMARY_FILE
+        fi
+        printf "python3 borealis_convert_file.py --dmap $(basename ${f})\n"
+        python3 "${HOME}/test_data_flow/data_flow/site-linux/borealis_convert_file.py" $f # For SAS testing
```
Test dir

RemingtonRohel commented on site-linux/convert_and_restructure:
```diff
@@ -0,0 +1,205 @@
+            printf "Removed records from ${f}:\n${remove_records_output}\n" | tee --append $SUMMARY_FILE
+        fi
+        printf "python3 borealis_convert_file.py --dmap $(basename ${f})\n"
+        python3 "${HOME}/test_data_flow/data_flow/site-linux/borealis_convert_file.py" $f # For SAS testing
```
Test dir

RemingtonRohel commented on site-linux/rsync_to_campus:
```diff
@@ -0,0 +1,145 @@
+readonly HDF5_SITES=("sas" "pgr" "inv" "cly" "rkn")
+
+# Location rawacf files are transferring from
+readonly DATA_DIR="/borealis_nfs/testing/inotify_data_flow" # For SAS testing
```
Test dir

RemingtonRohel commented on site-linux/rsync_to_campus:
```diff
@@ -0,0 +1,145 @@
+
+# If site isn't transferring dmap files, send to holding directory for campus conversion
+if [[ ! " ${DMAP_SITES[*]} " =~ " ${RADARID} " ]]; then
+	readonly DEST="/sddata/inotify_data_flow/${RADARID}_holding_dir" # For SAS testing
+else
+	readonly DEST="/sddata/inotify_data_flow/${RADARID}_data/" # For SAS testing
```
Test dirs

tjk584 commented on inotify_daemons/convert_on_campus.daemon:
```diff
@@ -21,6 +22,8 @@ do
-        $HOME/data_flow/superdarn-cssdp/convert_on_campus $RADAR_ID
+        touch $LOCK_FILE
+        # Lock with radar specific file to ensure only one instance of script runs for each site
+        flock -n --verbose $LOCK_FILE $HOME/data_flow/superdarn-cssdp/convert_on_campus $RADAR_ID
```
I've tested `flock` without using the -c argument, and I believe it works as well. From the man page for `flock`:

```
SYNOPSIS
       flock [options] file|directory command [arguments]
       flock [options] file|directory -c command
```

I did a quick test and verified the following commands are equivalent and both produce a lock correctly:

```
flock -n $LOCK_FILE ./test.sh $ARG1
flock -n $LOCK_FILE -c "./test.sh $ARG1"
```

The only difference I can see is that -c enforced a single argument as a command, so any executable using arguments would have to be contained within quotation marks.

kevinkrieger replied:
Ah! I misread the man page

RemingtonRohel commented on superdarn-cssdp/borealis_array_to_dmap.py:
```diff
@@ -168,10 +168,9 @@ def main():
-    print(time_now)
-    print(sys_call)
+    start_time = datetime.datetime.utcnow()
+    # sys_call = ' '.join(sys.argv[:])
+    # print(sys_call)
```
Can these be deleted?

tjk584 replied:
I believe so

RemingtonRohel commented on superdarn-cssdp/convert_on_campus:
```diff
@@ -0,0 +1,110 @@
+set -o pipefail  # don't hide errors within pipes
+
+# Load in function library
+source "${HOME}/test_data_flow/data_flow/library/data_flow_functions.sh"
```
Test dir

RemingtonRohel commented on superdarn-cssdp/convert_on_campus:
```diff
@@ -0,0 +1,110 @@
+    do
+        printf "\nConverting ${f}\n"
+        printf "python3 borealis_array_to_dmap.py $(basename ${f})\n"
+        python3 "${HOME}/data_flow/test_data_flow/superdarn-cssdp/borealis_array_to_dmap.py" $f # For SAS testing
```
Test dir

RemingtonRohel commented on superdarn-cssdp/distribute_borealis_data:
```diff
@@ -0,0 +1,171 @@
+set -o pipefail  # don't hide errors within pipes
+
+# Load in functions
+source "${HOME}/test_data_flow/data_flow/library/data_flow_functions.sh"
```
Test dir

RemingtonRohel commented on superdarn-cssdp/distribute_borealis_data:
```diff
@@ -0,0 +1,171 @@
+readonly RADAR_ID=$1
+
+# Define directories
+readonly TEST_DIR="${HOME}/test_data_flow" # For SAS testing
```
Test dir

tjk584 commented on inotify_daemons/convert_and_restructure.daemon:
```diff
@@ -8,6 +8,7 @@
 
 WATCH_DIR="${HOME}/data_flow/.inotify_watchdir"
 LOGFILE="${HOME}/logs/inotify_daemon.log"
+exec &>> $LOGFILE
```
Correct, I believe it redirects STDERR to LOGFILE as well. Pretty handy for organizing logs within a script

RemingtonRohel reviewed (changes requested):
Looks really good. There are a few places where testing directories are used, which I assume need to be changed if this is to be pulled on our computers. Once those are addressed I think it will be good to go.

kevinkrieger reviewed (approved):
I looked through the changes since Thursday, looks great

kevinkrieger reviewed (commented):
These review comments are from last week, I wasn't completed yet but I'll pass them to you now so you have them

tjk584 commented on borealis/rsync_to_nas:
```diff
@@ -0,0 +1,162 @@
+
+###################################################################################################
+
+readonly RADARID=$(hostname | cut --characters 1-3)
```
It's in the .bashrc for all site linux computers, but not the Borealis computers. We could change this though, to make it consistent across all computers.

tjk584 commented on borealis/rsync_to_nas:
```diff
@@ -0,0 +1,162 @@
+fi
+
+# Create log file. New file created daily
+readonly LOGGING_DIR="${HOME}/logs/rsync_to_nas/$(date +%Y)/$(date +%m)"
```
Nice, that looks a lot cleaner

tjk584 commented on borealis/rsync_to_nas:
```diff
@@ -0,0 +1,162 @@
+printf "Executing $0 on $(hostname)\n" | tee --append $SUMMARY_FILE
+date --utc "+%Y%m%d %H:%M:%S UTC" | tee --append $SUMMARY_FILE
+
+# Ensure that only a single instance of this script runs.
```
That's a good point, I think it would with the way I currently parse for summary data. I'll change this

tjk584 commented on site-linux/antennas_iq_backlog:
```diff
@@ -0,0 +1,112 @@
+
+# Specify which sites will convert each file type
+readonly ANTENNAS_IQ_SITES=("sas" "pgr" "inv" "cly" "rkn")
+readonly LOW_MEMORY_SITES=("pgr" "inv" "rkn")
```
Will do, must have missed that

tjk584 commented on inotify_daemons/borealis.daemon:
```diff
@@ -0,0 +1,78 @@
+exec &>> $LOGFILE # Redirect STDOUT and STDERR to $LOGFILE
+printf "Starting $0 at $(date --utc "+%Y%m%d %H:%M:%S UTC")\n"
+
+# Inotify is triggered multiple times as multiple 2-hour site files are created. We only want to 
```
For certain experiments, multiple rawacf/antennas_iq files are generated too, so the script would still trigger for each of those files

tjk584 commented on README.md:
```diff
@@ -1,21 +1,126 @@
+Chapman for uploading to the Engineering dashboard. SSH password-free connection must be setup 
+between each computer and Chapman for this to work correctly. 
+8. To modify the data flow easily, a `config.sh` file is provided. This file specifies:
+    - If the data flow can use the NAS at a site
```
Yes, I missed this for some reason

kevinkrieger reviewed (approved):
Was 15 seconds causing an issue somewhere? Just curious - changes look good to me

tjk584 commented:
> Was 15 seconds causing an issue somewhere? Just curious - changes look good to me

It caused the `rsync_to_nas` script to sometimes execute twice when 2f sound experiments ran. I don't think it caused any major problems though, at least none that showed up in the downtimes.
